cache/cache.go

```diff
@@ -1,95 +1,258 @@
 package cache
 
 import (
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/djherbis/atime"
+	"io"
+	"io/ioutil"
+	"log"
 	"os"
 	"path/filepath"
+	"sort"
 	"sync"
+	"net/http"
 )
 
-// Cache ...
+// ErrTooBig is returned by Cache::Put when when the item size is bigger than the
+// cache size limit.
+type ErrTooBig struct{}
+
+func (e *ErrTooBig) Error() string {
+	return "item bigger than the cache size limit"
+}
+
+// lruItem is the type of the values stored in SizedLRU to keep track of items.
+// It implements the SizedItem interface.
+type lruItem struct {
+	size      int64
+	committed bool
+}
+
+func (i *lruItem) Size() int64 {
+	return i.size
+}
+
+// Cache is the interface for a generic blob storage backend. Implementers should handle
+// locking internally.
 type Cache interface {
-	Dir() string
-	CurrSize() int64
+	// Put stores a stream of `size` bytes from `r` into the cache. If `expectedSha256` is
+	// not the empty string, and the contents don't match it, an error is returned
+	Put(key string, size int64, expectedSha256 string, r io.Reader) error
+	// Get writes the content of the cache item stored under `key` to `w`. If the item is
+	// not found, it returns ok = false.
+	Get(key string, w http.ResponseWriter) (ok bool, err error)
+	Contains(key string) (ok bool, err error)
+
+	// Stats
 	MaxSize() int64
-	AddFile(hash string, size int64)
-	RemoveFile(hash string) int64
-	ContainsFile(hash string) bool
+	CurrentSize() int64
 	NumFiles() int
-	LoadExistingFiles()
 }
 
-type cache struct {
-	dir      string
-	currSize int64
-	maxSize  int64
-	files    map[string]int64
-	mux      *sync.RWMutex
+// NewFsCache returns a new instance of a filesystem-based cache rooted at `dir`,
+// with a maximum size of `maxSizeBytes` bytes.
+func NewFsCache(dir string, maxSizeBytes int64) *fsCache {
+	// Create the directory structure
+	ensureDirExists(filepath.Join(dir, "cas"))
+	ensureDirExists(filepath.Join(dir, "ac"))
+
+	// The eviction callback deletes the file from disk.
+	onEvict := func(key Key, value SizedItem) {
+		// Only remove committed items (as temporary files have a different filename)
+		if value.(*lruItem).committed {
+			blobPath := filepath.Join(dir, key.(string))
+			err := os.Remove(blobPath)
+			if err != nil {
+				log.Println(err)
+			}
+		}
+	}
+
+	cache := &fsCache{
+		dir: dir,
+		mux: &sync.RWMutex{},
+		lru: NewSizedLRU(maxSizeBytes, onEvict),
+	}
+
+	cache.loadExistingFiles()
```

Comment:
Could we log here how many files we loaded into the cache?

Reply:
done